fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -4342,6 +5028,618 @@ struct RestoreLogDataTaskFunc : RestoreFileTaskFuncBase {
+			}
+		}
+
+		state std::vector<Reference<PartitionedLogIterator>> iterators(maxTagID + 1);
+		// for each tag, create an iterator
+		for (int k = 0; k < filesByTag.size(); k++) {
+			iterators[k] =
+			    makeReference<PartitionedLogIteratorTwoBuffers>(bc, k, filesByTag[k], fileEndVersionByTag[k]);
+		}
+
+		// mergeSort all iterator until all are exhausted
+		state int totalItereators = iterators.size();
+		// it stores all mutations for the next min version, in new format
+		state std::vector<Standalone<VectorRef<VersionedMutation>>> mutationsSingleVersion;
+		state bool atLeastOneIteratorHasNext = true;
+		state Version minVersion;
+		state int k;
+		state int versionRestored = 0;
+		state std::vector<Version> minVs(totalItereators, 0);
+		while (atLeastOneIteratorHasNext) {
+			minVs.resize(totalItereators, 0);
+			atLeastOneIteratorHasNext = false;
+			minVersion = std::numeric_limits<int64_t>::max();
+			k = 0;
+			for (; k < totalItereators; k++) {
+				if (!iterators[k]->hasNext()) {
+					continue;
+				}
+				atLeastOneIteratorHasNext = true;
+				Version v = wait(iterators[k]->peekNextVersion());
+				minVs[k] = v;
+
+				if (v <= minVersion) {
+					minVersion = v;
+				}
+			}
+			if (atLeastOneIteratorHasNext) {
+				k = 0;
+				for (; k < totalItereators; k++) {
+					if (!iterators[k]->hasNext()) {
+						continue;
+					}
+					Version v = wait(iterators[k]->peekNextVersion());
+					if (v == minVersion) {
+						Standalone<VectorRef<VersionedMutation>> tmp = wait(iterators[k]->getNext());
+						mutationsSingleVersion.push_back(tmp);
+					}
+				}
+
+				if (minVersion < begin) {
+					// skip generating mutations, because this is not within desired range
+					// this is already handled by the previous taskfunc
+					mutationsSingleVersion.clear();
+					continue;
+				} else if (minVersion >= end) {
+					// all valid data has been consumed
+					break;
+				}
+
+				// transform from new format to old format(param1, param2)
+				// in the current implementation, each version will trigger a mutation
+				// if each version data is too small, we might want to combine multiple versions
+				// for a single mutation
+				state Standalone<VectorRef<KeyValueRef>> oldFormatMutations =
+				    generateOldFormatMutations(minVersion, mutationsSingleVersion);
+				state int mutationIndex = 0;
+				state int txnCount = 0;
+				state int txBytes = 0;
+				state int totalMutation = oldFormatMutations.size();
+				state int txBytesLimit = CLIENT_KNOBS->RESTORE_WRITE_TX_SIZE;
+				state Key mutationLogPrefix = restore.mutationLogPrefix();
+
+				// multiple transactions are needed, so this has to be in a _execute method rather than
+				// a _finish method
+				loop {
+					try {
+						if (mutationIndex == totalMutation) {
+							break;
+						}
+						txBytes = 0;
+						txnCount = 0;
+						tr->reset();
+						tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+						tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+
```

Comment:
Might be a good idea to add `wait(taskBucket->keepRunning(tr, task));` here, or explain that this transaction does blind writes, so they are idempotent operations even if multiple instances of the same tasks are running these transactions.